apps/src/studio/studio.js

```diff
@@ -5858,6 +5868,54 @@ Studio.getSkin = function () {
 };
 
 /**
+ * For grid-aligned movement, we want a single movement action to take place
+ * over several ticks (as opposed to normal movement, which takes place on a
+ * per-tick basis). We therefore yield control for the movement duration.
+ *
+ * @see Studio.turnSingle
+ * @see Studio.moveSingle
+ */
+Studio.yieldGridAlignedTicks = function () {
+  Studio.yieldExecutionTicks += (1 + Studio.gridAlignedExtraPauseSteps);
+  if (Studio.JSInterpreter) {
+    // Stop executing the interpreter in a tight loop and yield the current
+    // execution tick:
+    Studio.JSInterpreter.yield();
+    // Highlight the code in the editor so the student can see the progress
+    // of their program:
+    Studio.JSInterpreter.selectCurrentCode();
+  }
+
+  Studio.movementAudioOn();
+};
+
+/**
+ * For executing a single "goLeft" or "goNorth" sort of command in student code.
+ * Moves the avatar by a different amount.
+ * Has slightly different behaviors depending on whether the level is configured
+ * for discrete, grid-based movement or free movement.
+ * @param {Object} opts
+ * @param {Direction} opts.dir - The direction in which the sprite should move.
+ * @param {number} opts.spriteIndex
+ * @param {boolean} opts.backward - whether the sprite should move toward
+ *        (default) or away from the given direction
+ */
+Studio.turnSingle = function (opts) {
+  const sprite = Studio.sprite[opts.spriteIndex];
+  sprite.lastMove = Studio.tickCount;
+
+  if (skin.gridAlignedMovement) {
+    sprite.setActivity(constants.BEHAVIOR_GRID_ALIGNED);
+    sprite.addAction(new GridTurn(opts.dir, skin.slowExecutionFactor));
+    Studio.yieldGridAlignedTicks();
+  } else {
+    throw new TypeError("Studio.turnSingle is only valid in grid-aligned mode");
```

Comment:
Style nit: You could check this condition and throw as an early-out at the top of this method, which would make it more obvious that the rest of the method is a simple command list with no branches.

```js
Studio.turnSingle = function (opts) {
  if (!skin.gridAlignedMovement) {
    throw new TypeError("Studio.turnSingle is only valid in grid-aligned mode");
  }

  const sprite = Studio.sprite[opts.spriteIndex];
  sprite.lastMove = Studio.tickCount;
  sprite.setActivity(constants.BEHAVIOR_GRID_ALIGNED);
  sprite.addAction(new GridTurn(opts.dir, skin.slowExecutionFactor));
  Studio.yieldGridAlignedTicks();
  Studio.lastMoveSingleDir = opts.dir;
};
```